worker/changestreampruner/worker.go

```diff
@@ -0,0 +1,396 @@
+		return Watermark{}, errors.Trace(err)
+	}
+
+	// Nothing to do if there are no watermarks.
+	if len(watermarks) == 0 {
+		return Watermark{}, nil
+	}
+
+	// Gather all the watermarks that are within the windowed time period.
+	// If there are no watermarks within the window, then we can assume
+	// that the stream is keeping up and we don't need to prune anything.
+	lowest, ok := lowestWatermark(watermarks, w.cfg.Clock.Now())
+	if !ok {
+		// Check to see if the latest change log has a valid log in the last
+		// window duration, if not, then we can assume that the stream is not
+		// keeping up and we should log a warning.
+		var changes []ChangeLog
+		if err := tx.Query(ctx, selectChangeLogQuery, sqlair.M{
+			"created_at": w.cfg.Clock.Now().Add(-defaultWindowDuration),
+			"limit":      changestream.DefaultNumTermWatermarks + 1,
+		}).GetAll(&changes); err != nil {
+			return Watermark{}, errors.Trace(err)
+		}
+		// If there are less than the default number of term watermarks, then
+		// we can assume that the stream is keeping up and we don't need to
+		// prune anything.
+		if len(changes) < changestream.DefaultNumTermWatermarks {
+			return Watermark{}, nil
+		}
+		w.cfg.Logger.Infof("No watermarks within window, check logs to see if the change stream is keeping up")
+		return Watermark{}, nil
+	}
+	return lowest, nil
+}
+
+var pruneQuery = sqlair.MustPrepare(`DELETE FROM change_log_witness WHERE controller_id NOT IN (SELECT controller_id FROM controller_node);`)
+
+func (w *Pruner) pruneWitnessTable(ctx context.Context, tx *sqlair.TX) error {
+	// Prune the change log witness table, this ensures that we don't have
+	// any dangling watermarks. As the controller_node table is HA aware,
+	// which is kept up to date, we can use that to remove any rows from
+	// the change_log_witness table that are no longer valid.
+	if err := tx.Query(ctx, pruneQuery).Run(); err != nil {
+		return errors.Trace(err)
+	}
+	return nil
+}
+
+var deleteQuery = sqlair.MustPrepare(`DELETE FROM change_log WHERE id <= $M.id;`, sqlair.M{})
+
+func (w *Pruner) deleteChangeLog(ctx context.Context, tx *sqlair.TX, lowest Watermark) (int64, error) {
+	// Delete all the change logs that are lower than the lowest watermark.
+	var outcome sqlair.Outcome
+	if err := tx.Query(ctx, deleteQuery, sqlair.M{"id": lowest.LowerBound}).Get(&outcome); err != nil {
+		return -1, errors.Trace(err)
+	}
+	pruned, err := outcome.Result().RowsAffected()
+	return pruned, errors.Trace(err)
+}
+
+func lowestWatermark(watermarks []Watermark, now time.Time) (Watermark, bool) {
+	// Select the lower bound of the watermark, only if the updated_at time
+	// is within a windowed time period.
+	var (
+		view = window{
+			start: now.Add(-defaultWindowDuration),
+			end:   now,
+		}
+		lowest = Watermark{
+			LowerBound: math.MaxInt64,
+		}
+	)
+	for _, watermark := range watermarks {
+		// If the watermark is outside of the window, skip it.
+		if !view.contains(watermark.UpdatedAt) {
```

Comment:
This doesn't seem right right. 

If I have watermark 2 earlier than the window, this is discounted as a candidate for lowest.
If the other 2 controllers have 4, and 5 within the window, then 4 is chosen, and we delete 3, which one controller has not acknowledged. Is that right?

Reply:
We decided to log a warning instead of ignoring a watermark. This will give us time in the field to check if this strategy is a valid one. In reality, it shouldn't happen, but we can't be 100% sure it won't.